Global Styles: Add support for caption elements

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This makes it possible to style captions in theme.json by adding this into your theme.json file:

```
{
	"styles": {
		"elements": {
			"caption": {
				"color": {
					"background": "red",
					"text": "yellow"
				}
			}
		}
	}
}
```

## Why?
So that themes can independently control the display of their captions. See https://github.com/WordPress/gutenberg/issues/40205

## How?
We create a new element called caption and then add the `wp-element-caption` class to all elements that we want to use the caption styles.

The problem with this approach is that captions are also targeted by the default styles in the block themselves. This is not a problem as the CSS global styles outputs is stronger, except for the case of captions inside gallery blocks.

## Testing Instructions
- Add the snippet above to your theme
- Add an audio/image/table/embed/video/gallery block to the page
- Confirm that the captions have the styles added above.

## Screenshots or screencast <!-- if applicable -->
<img width="1415" alt="Screenshot 2022-06-16 at 14 00 22" src="https://user-images.githubusercontent.com/275961/174075248-3f1e0428-fe71-4290-b7ab-5c97553f42ae.png">

carolinan commented:
If we choose this path, does it mean that we are no longer considering making the captions separate inner blocks?

scruffian commented:
> If we choose this path, does it mean that we are no longer considering making the captions separate inner blocks?

I believe so yes. Inner blocks bring a lot of extra complexity with them (for example the block toolbar and inspector controls) which may not give us the best user experience for editing the block they are part of.

scruffian commented:
cc @WordPress/block-themers

draganescu reviewed (approved):
I believe this follows in the footsteps of adding the button element. 🚢 it.

@carolinan it's not known. We're adding element support so that `theme.json` can style this element, but if we arrive at a place where a `Caption` block makes sense it will not be blocked by the element, but it will probably use it and inherit the styling seamlessly.

scruffian committed:
dd caption element

dcalhoun committed:
test: Revert image caption class name addition for native test fixtures

The current save output for native editor Image and Gallery blocks do
not expect this class name. I did not identify the exact cause, but this
does fix the failing existing tests.

dcalhoun committed:
test: Reinstate image caption class name for Image block test fixture

The native editor test fixtures includes a Gallery block v1 (which needs
to upgraded to v2 at some point). Adding a `wp-element-caption` to this
test fixture causes parsing failure.

The lack of `wp-element-caption` class name for the Image block causes
an unnecessary block update during test runs.

The `wp-element-caption` class name was added to the native editor test
fixtures for both the Gallery and Image. The addition was only needed
for the latter. The previous commit erroneously removed the class name
for both. This commit reinstates the class name for the latter.

scruffian committed:
update snapshot

ndiego commented:
@scruffian, this seems to work in the Editor, but not 100% on the Frontend. In my testing, certain values are overridden by certain block styles: 
![image](https://user-images.githubusercontent.com/4832319/177624414-3e01e75e-81a7-4274-a7c7-6cfacb8a340d.png)

bgardner commented:
@scruffian +1 to @ndiego's comment. Seems similar to this: https://github.com/WordPress/gutenberg/pull/41934

scruffian commented:
We have an issue for this: https://github.com/WordPress/gutenberg/issues/42054

mburridge commented:
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.

scruffian commented:
I think this is covered by: https://make.wordpress.org/core/?p=98557

mburridge commented:
Thanks @scruffian, I've removed the label.

youknowriad commented on packages/block-library/src/audio/save.js:
```diff
@@ -17,7 +21,13 @@ export default function save( { attributes } ) {
+					<RichText.Content
+						tagName="figcaption"
+						value={ caption }
+						className={ __experimentalGetElementClassName(
```
Any change to a save function of a block need a deprecation to be aded to that block. See https://github.com/WordPress/gutenberg/issues/45153
